### PR TITLE
replaced the range with the control knob component

### PR DESCRIPTION
**Pull request by nspilman**
- installed react-dial-knob
- created controlKnob component
- deleted controlRange and replaced it with controlKnob

### CODE REVIEW

File: src/components/controls/components/dialControl.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    const { min, 
+            max, 
+            localStorageKey, 
+            title, 
+            defaultValue, 
+            factor } = props.parameters;
+    
+    const { setValue } = props;
+                
+    let initialOscCount = localStorageService.getNumberByKey(localStorageKey) ?? defaultValue;
```
Comment (soiferj) [inline src/components/controls/components/dialControl.tsx]: `initialState`?

File: src/components/controls/basic/waveControl.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+function WaveControl(){
+    const state : IKeyboardContextSignature = useContext(KeyboardContext);
+    const setState = useContext(UpdateKeyboardContext);
+    const localStorageKey = "waveform";
+    const waveTypeOptions = getAllWaveTypes();
+  
+    const { audioContextWrapper }: IKeyboardContextSignature = useContext(KeyboardContext);
+    let initialWaveform = localStorage.getItem(localStorageKey) ?? waveTypeOptions[0];
+
+    const [wave, setWave] = useState(initialWaveform);
```
Comment (soiferj) [inline src/components/controls/basic/waveControl.tsx:26]: Can you explain why we need both `useState` and `useContext`? I thought the state was included when we got the context?

File: src/components/controls/envelopeControlsWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+const StyledEnvelopeControl = styled.div`
+    display:flex;
+    align-items:center;
+    justify-content:center;
+`
+
+function EnvelopeControl(){
+   const { audioContextWrapper } :IKeyboardContextSignature = useContext(KeyboardContext);
+    console.log({audioContextWrapper})
```
Comment (soiferj) [inline src/components/controls/envelopeControlsWrapper.tsx]: Remove?

File: src/components/controls/envelopeControlsWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            />
+            <DialControl 
+                parameters = {envelopeDecayParameters}
+                setValue = {(value) => audioContextWrapper.setDecayMs(value) }
+            />
+            <DialControl 
+                parameters = {envelopeReleaseParameters}
+                setValue = {(value) => audioContextWrapper.setReleaseMs(value)}
+            />
+            <DialControl
```
Comment (soiferj) [inline src/components/controls/envelopeControlsWrapper.tsx]: Move sustain before release?

File: src/components/controls/oscillator/oscillatorCountControl.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+function OscillatorCountControl(){
+    const localStorageKey = 'numOscillators';
+    const { audioContextWrapper }: IKeyboardContextSignature = useContext(KeyboardContext);
+    let initialOscCount = localStorageService.getNumberByKey(localStorageKey) ?? 1;
+    const [numOscillators, setNumOscillators] = useState(initialOscCount);
+    const minNumOscillators = 1; 
+    const maxNumOscillators = 2; 
+
+    useEffect(() : void=> {
+        audioContextWrapper.setNumOscillators(numOscillators);
```
Comment (soiferj) [inline src/components/controls/oscillator/oscillatorCountControl.tsx:18]: I sent this over text too, but I feel that the usage of of context, state, and storage is a little confusing. Can any of it be condensed? Or moved into a common file to set everything together?

File: src/data/dialControlParmeters.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    1,
+    2000,
+    "DECAY",
+    "envelopeDecayMs",
+    defaultValues.envelopeParameters.decayMs,
+) 
+
+export const envelopeSustainParameters  = createParametersObject(
+    0,
+    1000,
```
Comment (soiferj) [inline src/data/dialControlParmeters.tsx:41]: Why is this 0 to 1000? Sustain can technically only go from 0 to 1.

File: src/data/dialControlParmeters.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    40,
+    20000,
+    "FREQUENCY",
+    "filterFrequency",
+    defaultValues.filterParameters.freq
+)
+
+export const filterQParameters = createParametersObject(
+    1,
+    20000,
```
Comment (soiferj) [inline src/data/dialControlParmeters.tsx]: Q should go from 0.01 to 20 (these were the previous bounds too)

File: src/services/localStorageService.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+const localStorageService : ILocalStorageService = {
+    getNumberByKey : (key : string) => {
+        const value = localStorage.getItem(key);
+        if(value){
+            return JSON.parse(value);
+        }
+    },
+    setValueByKey : (key : string, value :string) => {
+        localStorage.setItem(key,value);
```
Comment (soiferj) [inline src/services/localStorageService.tsx:11]: Why does this not need to be JSONified?

File: src/components/controls/components/dialControl.tsx (hunk 1 above)
Comment (nspilman) [reply src/components/controls/components/dialControl.tsx]: Yes - good catch.

File: src/components/controls/basic/waveControl.tsx (hunk 2 above)
Comment (nspilman) [reply src/components/controls/basic/waveControl.tsx:26]: useState keeps track of the state within the component. UseContext is used to have access to the global audioContextWrapper

File: src/components/controls/envelopeControlsWrapper.tsx (hunk 3 above)
Comment (nspilman) [reply src/components/controls/envelopeControlsWrapper.tsx]: Yessir

File: src/components/controls/oscillator/oscillatorCountControl.tsx (hunk 5 above)
Comment (nspilman) [reply src/components/controls/oscillator/oscillatorCountControl.tsx:18]: Responded over text, but no harm in too my documentation. useState keeps track of state within a component. useContext gives the components access to the global audioContextParameter. useEffect calls the audioContextWrapper method to set the value every time the useState value changes.

File: src/services/localStorageService.tsx (hunk 8 above)
Comment (nspilman) [reply src/services/localStorageService.tsx:11]: setItem handles it under the hood

File: src/data/dialControlParmeters.tsx (hunk 7 above)
Comment (nspilman) [reply src/data/dialControlParmeters.tsx]: I haven't figured out how to get the knob to play nice with non-integers

File: src/data/dialControlParmeters.tsx (hunk 6 above)
Comment (nspilman) [reply src/data/dialControlParmeters.tsx:41]: I haven't figured out how to get the knob to play nice with non-integers

**Pull request merged**